Add support for native cs type parameter constraints

kLabz opened the pull request:
Allow one to define native C# constraints from `haxe.Constraints.Constructible` or `cs.Constraints.*`. This is useful when using C# APIs using such constraints for inputs; without it we need "ugly" workarounds.

Supported native constraints:
* `where T : struct` with `cs.Constraints.CsStruct`
* `where T : class` with `cs.Constraints.CsClass`
* `where T : unmanaged` with `cs.Constraints.CsUnmanaged`
* `where T : new()` with `haxe.Constraints.Constructible<Void->Void>`

For example, this:
```haxe
import cs.Constraints;

@:nativeGen
class Main<T:CsStruct> {
	public function new() {}
}
```

Will result in:
```cs
public class Main<T> where T : struct {
	public Main() {
	}
}
```

See Microsoft's [documentation on the subject](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters) for reference.

kLabz committed:
Add support for native cs type parameter constraints

Gama11 commented:
Is this a case where `@:coreType` would make sense to avoid the abstract needing a (`Dynamic`) underlying type?

kLabz commented:
Could be; although `from Dynamic` (or something nicer) would still be needed.

RealyUniqueName commented on std/cs/Constraints.hx:
```diff
@@ -0,0 +1,53 @@
+	using `@:nativeGen`. This constraint won't have any effect on Haxe code.
+	If used as a real type, the underlying type will be `Dynamic`.
+**/
+abstract IsClass(Dynamic) from Dynamic {}
```
Maybe rename to `CsStruct`, `CsClass` and `CsUnmanaged`?
Because `IsClass` does not match the naming of others. 
And it will reflect relevant symbols on the ocaml side of Haxe.

kLabz replied:
Good point, I'll do that.

Gama11 replied:
Hm, but having a "hungarian notation style" prefix for everything is ugly too. :P

Is it really a problem to use `cs.Constraints.Class`? Especially if its used in a fully qualified manner, it shouldn't really cause conflicts.

RealyUniqueName replied:
It will shadow top level `Class<T>` if imported.

RealyUniqueName replied:
And cause confision. Because everyone, who see the usage of that `Class` in code will think about Haxe's `Class<T>` and not about `cs.Constraints.Class`

Gama11 replied:
I know, but is that really a problem in practice? We have enum constructors from various different enums named `Class` / `String` in haxe-language-server for instance, and I've never really had issues with it. You can still prefix `std` to fix conflicts.

kLabz committed:
Unify native C# Constraints names

RealyUniqueName commented:
Add a test, please.

kLabz committed:
Cs new() constraint: only for Constructible<Void->Void>

kLabz committed:
Cs constraints: class and struct are incompatible too

kLabz committed:
Cs constraints: unmanaged is only available with cs >= 7.3

kLabz committed:
Cs native constraints: add tests

Gama11 commented:
`cs_ver` should probably be added to `define.ml` so it shows up in `haxe --help-defines`?

kLabz committed:
Add cs_ver to define.ml

kLabz commented:
Notes on [tests](https://github.com/HaxeFoundation/haxe/pull/8311/commits/74d2f4627581182a3d3eb5202d082a0a548a591a):

* `TestCs` class may seem redundant but is actually required to make sure the generated `where` constraints are present in the generated C# code so that we can use C# APIs requiring them.

* Both static functions and classes are tested in `Main.hx`, again it seems to be redundant but it makes sure native constraints work in both places

* Unsupported combinations as [documented on MSDN](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters) are tested in `incompatible-combinations-fail.hxml`. This includes the undocumented incompatibility between `class` and `struct` (both have to be the very first constraint).

kLabz committed:
Fix copy/paste fail and add corresponding test

kLabz commented:
Should add 
```json
	{
		"name": "CsVer",
		"define": "cs_ver",
		"doc": "The C# version to target",
		"platforms": ["cs"]
	},
```
in #8195 if this is merged.

Edit: #8195 has been merged, and this PR has been updated.

kLabz commented:
@Gama11 
> Is this a case where @:coreType would make sense to avoid the abstract needing a (Dynamic) underlying type?

Should I do this? I used `haxe.Constraints` as a model for `cs.Constraints`, but I can change it (and maybe update `haxe.Constraints` too?).

Gama11 commented:
I'm not sure, is there a reason why `haxe.Constraints` types are defined the way they are @Simn?

kLabz committed:
Use @:coreType in cs.Constraints

kLabz commented:
> simn
> `@:coreType` typically has a specific native representation which is handled by the target generator

> Rudy G
> Would you say then that `@:coreType` would make more sense in cs.Constraints since they generate native c# type parameter constraints?
> (and not necessarily in haxe.Constraints)

> simn
> Yes if these typed are picked up by the generator you can use `@:coreType`
> Abstracts over `Dynamic` are for when the run-time representation is not specified.

I updated `cs.Constraints` to use `@:coreType` (and not `haxe.Constraints`) and rebased development again, waiting for CI ~
